Description: Needed for kfreebsd-* arch
 Fix link_settings.
Forwarded: not-needed
Author: Steven Chamberlain <steven@pyro.eu.org>
Author: Jérémy Lal <kapouer@melix.org>
Bug-Debian: http://bugs.debian.org/670836
Last-Update: 2012-10-05
--- a/tools/gyp/v8.gyp
+++ b/tools/gyp/v8.gyp
@@ -667,7 +667,7 @@
             ['OS=="freebsd"', {
                 'link_settings': {
                   'libraries': [
-                    '-L/usr/local/lib -lexecinfo',
+                    '-L/usr/local/lib -lpthread',
                 ]},
                 'sources': [
                   '../../src/platform-freebsd.cc',
