Description: Fix armhf (again)
Bug-Debian: http://bugs.debian.org/636370
Author: Konstantinos Margaritis <markos@genesi-usa.com>
Last-Update: 2011-08-24

--- a/SConstruct
+++ b/SConstruct
@@ -885,6 +885,10 @@
   return 'default'
 
 
+def GuessARMEABI(env):
+  eabi = env['armeabi'];
+  return eabi
+
 PLATFORM_OPTIONS = {
   'arch': {
     'values': ['arm', 'ia32', 'x64', 'mips'],
@@ -900,7 +904,12 @@
     'values': ['gcc', 'msvc'],
     'guess': GuessToolchain,
     'help': 'the toolchain to use'
-  }
+  },
+  'armeabi': {
+    'values': ['hard', 'softfp', 'soft'],
+    'default': 'softfp',
+    'help': 'generate calling conventiont according to selected ARM EABI variant'
+  },
 }
 
 SIMPLE_OPTIONS = {
@@ -1020,11 +1029,11 @@
     'default': 'off',
     'help': 'select profile guided optimization variant',
   },
-  'armeabi': {
-    'values': ['hard', 'softfp', 'soft'],
-    'default': 'softfp',
-    'help': 'generate calling conventiont according to selected ARM EABI variant'
-  },
+#  'armeabi': {
+#    'values': ['hard', 'softfp', 'soft'],
+#    'guess': GuessARMEABI,
+#    'help': 'generate calling conventiont according to selected ARM EABI variant'
+#  },
   'mipsabi': {
     'values': ['hardfloat', 'softfloat', 'none'],
     'default': 'hardfloat',
@@ -1165,6 +1174,8 @@
     Abort("Option unalignedaccesses only supported for the ARM architecture.")
   if env['os'] != 'linux' and env['compress_startup_data'] != 'off':
     Abort("Startup data compression is only available on Linux")
+  if (env['arch'] == 'arm' and not (env['armeabi'] in ('hard', 'softfp', 'soft'))):
+    Abort("Invalid EABI set for arm, use one of hard, softfp, soft.");
   for (name, option) in ALL_OPTIONS.iteritems():
     if (not name in env):
       message = ("A value for option %s must be specified (%s)." %
