Description: build samples with gyp and system libv8
Forwarded: not-needed
Author: Jérémy Lal <kapouer@melix.org>
Last-Update: 2011-12-05
--- a/samples/samples.gyp
+++ b/samples/samples.gyp
@@ -26,15 +26,20 @@
 # OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 {
-  'includes': ['../build/common.gypi'],
+  'includes': [],
   'target_defaults': {
     'type': 'executable',
     'dependencies': [
-      '../tools/gyp/v8.gyp:v8',
+      
     ],
     'include_dirs': [
-      '../include',
+     '/usr/include'
     ],
+    'link_settings': {
+	   'libraries': [
+	      '-lv8',
+	   ],
+	},
   },
   'targets': [
     {
@@ -48,6 +53,12 @@
       'sources': [
         'process.cc',
       ],
+    },
+    {
+      'target_name': 'lineprocessor',
+      'sources': [
+        'lineprocessor.cc',
+      ],
     }
   ],
 }
