Description: fix "make check" for mips arch.
Forwarded: not-needed, upstream already fixed this in 3.8 branch.
Author: Jérémy Lal <kapouer@melix.org>
Last-Update: 2012-02-20
--- a/tools/test-wrapper-gypbuild.py
+++ b/tools/test-wrapper-gypbuild.py
@@ -146,7 +146,7 @@
       print "Unknown mode %s" % mode
       return False
   for arch in options.arch:
-    if not arch in ['ia32', 'x64', 'arm']:
+    if not arch in ['ia32', 'x64', 'arm', 'mips']:
       print "Unknown architecture %s" % arch
       return False
 
--- a/test/mjsunit/mjsunit.status
+++ b/test/mjsunit/mjsunit.status
@@ -122,6 +122,9 @@
 
 ##############################################################################
 [ $arch == mips ]
+# Failing for now but not crucial
+debug-step-3: PASS || FAIL
+
 # Run those tests, but expect them to time out.
 array-sort: PASS || TIMEOUT
 mirror-object: PASS || TIMEOUT
